### PR TITLE
Overhaul Documenting Code

### PR DESCRIPTION
**Pull request by straight-shoota**
This is a major overhaul of the *Documenting Code* page.

* adds structural sub-sections
* rewrites the content more as a specification than an informal guide
* extends missing documentation
* improves examples

The entire section about the formatter is replaced by a reference to the compiler manual (and an note there that the formatter also applies to code blocks in doc comments).

### CODE REVIEW

File: docs/conventions/documenting_code.md [hunk 1]
```diff
@@ -1,119 +1,210 @@
...
-    ```crystal
-    class Unicorn
-      # :nodoc:
-      class Helper
-      end
-    end
-    ```
+The compiler implicitly adds some admonitions to doc comments:
+
+* The [`@[Deprecated]`](https://crystal-lang.org/api/Deprecated.html) annotation
```
Comment (bcardiff) [inline docs/conventions/documenting_code.md:159]:
```suggestion
* The [`@[Deprecated]`](https://crystal-lang.org/api/latest/Deprecated.html) annotation
```

File: docs/conventions/documenting_code.md [hunk 2]
```diff
@@ -1,119 +1,210 @@
...
+  adds a `DEPRECATED` admonition.
+* The [`@[Experimental]`](https://crystal-lang.org/api/Experimental.html) annotation
```
Comment (bcardiff) [inline docs/conventions/documenting_code.md:161]:
```suggestion
* The [`@[Experimental]`](https://crystal-lang.org/api/latest/Experimental.html) annotation
```

File: docs/using_the_compiler/README.md [hunk 3]
```diff
@@ -235,7 +235,7 @@ Initialized empty Git repository in ~/my_cool_lib/.git/
 
 ### `crystal docs`
 
-The `crystal docs` command generates API documentation from inline docstrings in Crystal files (see [documenting code](../conventions/documenting_code.md)).
+The `crystal docs` command generates API documentation from inline docstrings in Crystal files (see [*Documenting Code*](../conventions/documenting_code.md)).
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply docs/using_the_compiler/README.md:238]: No, not in general. It's a reference to a section title. And in in my style guide reference to a title are written in italics (and in the same casing). =)

File: docs/conventions/documenting_code.md (hunk 1 above)
Comment (straight-shoota) [reply docs/conventions/documenting_code.md:159]: Most links on other pages don't include `latest` in the path. Do we want to standardize on either form? Is there a specific benefit of using `latest`?

Comment (bcardiff) [reply docs/conventions/documenting_code.md:159]: One less redirect / request.
The urls without /latest/ were the original location before we split it in versions.

File: docs/using_the_compiler/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bcardiff) [reply docs/using_the_compiler/README.md:238]: ok

**Review by bcardiff — APPROVED** [review]

File: docs/conventions/documenting_code.md [hunk 4]
```diff
@@ -1,119 +1,210 @@
...
-    # described since antiquity as a beast with a large, spiraling horn projecting
-    # from its forehead.
-    class Unicorn
-    end
+## Association
 
-    # Bad: This is not attached to any class.
+Doc comments must be positioned directly above the definition of the
+documented feature. Consecutive comment lines are combined into a single comment
+block. Any empty line breaks the association to the documented feature
```
Comment (vlazar) [inline docs/conventions/documenting_code.md:18]:
```suggestion
block. Any empty line breaks the association to the documented feature.
```

**Review by sdogruyol — APPROVED** [review]
Thank you @straight-shoota 🙏

**Comment by vlazar** [comment]
As a non native speaker I had to check dictionary for "admonition". Can it be replaced with someg more frequently used word to make docs easier to read?

**Comment by straight-shoota** [comment]
It's hard to find an approriate alternative. The previous version described it as "flagging", but that term is very imprecise and already overloaded with several semantics.
*admonition* is used as a technical term and as such it can hardly be replaced. We could improve by describing the word for readers unfamiliar with it.

**Pull request merged**